teuthology/suite/test/test_run_.py

```diff
@@ -232,9 +233,10 @@ def test_successful_schedule(
         m_build_matrix.return_value = build_matrix_output
         frag1_read_output = 'field1: val1'
         frag2_read_output = 'field2: val2'
-        m_file.side_effect = [
+        m_open.side_effect = [
             StringIO(frag1_read_output),
             StringIO(frag2_read_output),
+            contextlib.closing(StringIO())
```

Comment:
What's this do?

Reply:
in short it makes the test pass

Reply:
https://docs.python.org/3/library/contextlib.html#contextlib.closing

so it's for automatically closing resources opened with the ContextManager, which makes sense, but I don't get how passing it a StringIO() works. I guess it closes all opened StringIO objects somehow? This level of python-foo is beyond me and google isn't helping. :(

Reply:
The side_effect array is used to replace output of each call, in this particular case it is open() method. The call of this method is found three times in this particular case:
Two of them are called via this route:
```
test_successful_schedule -> schedule_suite -> collect_jobs
```
where two fragments are read in this code:
```
    raw_yaml = '\n'.join([open(a, 'r').read() for a in fragment_paths])
```
And the third one is via this route:
```
test_successful_schedule -> schedule_suite
```
in the block:
```
        with open(base_yaml_path, 'w+b') as base_yaml:
            base_yaml.write(str(self.base_config))
```
We've been using a `file()` class before, which is not a context type and has no `__exit__` method. But after we switched to `open()` function, which return an context object with `__exit__` method, then on exit from `with ...` block it is invoked (for the file() python interpreter just didn't call it. This why we need to add extra side_effect with value of an object with `__exit__` method additionally declared, like based on contextlib.closing() class. The open() function also returns an object with `close()` method which is used as well, this is why we need to use something like StringIO() class, etc.

I hope this helps.